kythe/go/serving/xrefs/columnar.go

```diff
@@ -0,0 +1,242 @@
+
+	// Setup scanning state for constructing reply
+	var norm *span.Normalizer                                          // span normalizer for references
+	refsByTarget := make(map[string][]*xpb.DecorationsReply_Reference) // target -> set<Reference>
+	defs := stringset.New()                                            // set<needed definition tickets>
+	patterns := xrefs.ConvertFilters(req.Filter)
+	emitSnippets := req.Snippets != xpb.SnippetsKind_NONE
+
+	// Main loop to scan over each columnar kv entry.
+	for {
+		k, val, err := it.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		key := string(k[len(prefix):])
+
+		// TODO(schroederc): only parse needed entries
+		e, err := columnar.DecodeDecorationsEntry(file, key, val)
+		if err != nil {
+			return nil, err
+		}
+
+		switch e := e.Entry.(type) {
+		case *xspb.FileDecorations_Text_:
+			file := e.Text
+			norm = span.NewNormalizer(file.Text)
+
+			loc, err := norm.Location(req.GetLocation())
+			if err != nil {
+				return nil, status.Errorf(codes.InvalidArgument, "invalid Location: %v", err)
+			}
+			reply.Location = loc
+
+			if req.SourceText {
+				reply.Encoding = idx.TextEncoding
+				if loc.Kind == xpb.Location_FILE {
+					reply.SourceText = file.Text
+				} else {
+					reply.SourceText = file.Text[loc.Span.Start.ByteOffset:loc.Span.End.ByteOffset]
+				}
+			}
+		case *xspb.FileDecorations_Target_:
+			if !req.References {
+				// TODO(schroederc): seek to next group
+				continue
+			}
+			t := e.Target
+			kind := t.GetGenericKind()
+			if kind == "" {
+				kind = schema.EdgeKindString(t.GetKytheKind())
+			}
+			ref := &xpb.DecorationsReply_Reference{
+				TargetTicket: kytheuri.ToString(t.Target),
+				Kind:         kind,
+				Span:         norm.SpanOffsets(t.StartOffset, t.EndOffset),
+			}
+			refsByTarget[ref.TargetTicket] = append(refsByTarget[ref.TargetTicket], ref)
+			reply.Reference = append(reply.Reference, ref)
+		case *xspb.FileDecorations_TargetOverride_:
+			// TODO(schroederc): handle
+		case *xspb.FileDecorations_TargetNode_:
+			if len(patterns) == 0 {
+				// TODO(schroederc): seek to next group
+				continue
+			}
+			n := e.TargetNode.Node
+
+			c := &cpb.NodeInfo{Facts: make(map[string][]byte, len(n.Fact))}
+			for _, f := range n.Fact {
+				name := schema.GetFactName(f)
+				if xrefs.MatchesAny(name, patterns) {
+					c.Facts[name] = f.Value
+				}
+			}
+			if kind := schema.GetNodeKind(n); kind != "" && xrefs.MatchesAny(facts.NodeKind, patterns) {
+				c.Facts[facts.NodeKind] = []byte(kind)
+			}
+			if subkind := schema.GetSubkind(n); subkind != "" && xrefs.MatchesAny(facts.Subkind, patterns) {
+				c.Facts[facts.Subkind] = []byte(subkind)
+			}
+			if len(c.Facts) > 0 {
+				reply.Nodes[kytheuri.ToString(n.Source)] = c
+			}
+		case *xspb.FileDecorations_TargetDefinition_:
+			if !req.TargetDefinitions {
+				continue
+			}
+			def := e.TargetDefinition
+			refs := refsByTarget[kytheuri.ToString(def.Target)]
```

Comment:
Is refsByTarget always going to be filled by the time we get to TargetDefinitions in the table?